src/vsphere_cpi/lib/cloud/vsphere/storage_picker.rb

```diff
@@ -53,25 +53,56 @@ def choose_ephemeral_storage(target_datastore_name, accessible_datastores, vm_ty
       choose_best_from(storage_options)
     end
 
-    # Ephemeral Pattern is constructed from given datastores and collection of all datastores for sdrs enabled datastore clusters
-    # if no datastores/datastore_clusters are specified, use global ephemeral pattern
-
+    # Ephemeral Pattern is constructed using following list of priority
+    # 1. vm_type storage policy
+    # 2. vm_type ephemeral DS pattern
+    # 3. Global Storage Policy
+    # 4. Global ephemeral DS pattern
+    #
+    # if policy is specified used compatible datastores for that policy otherwise
+    # use given datastores and collection of all datastores for sdrs enabled datastore clusters
+    # and if no datastores/datastore_clusters are specified, use global ephemeral pattern
+    #
     # @param [VmType] vm_type
-    def choose_ephemeral_pattern(vm_type)
+    # @param [Config] global_config
+    #
+    # @return [String]DS Pattern, [String, nil] PolicyName (if applicable)
+    def choose_ephemeral_pattern(global_config, vm_type)
+
+      # 1. Check for vm_type storage policy first
+      # return if exists
+      if vm_type.storage_policy_name
+        datastore_names = vm_type.storage_policy_datastores(vm_type.storage_policy_name).map{|d| d.name }
+        logger.info("Using compatible datastores: #{datastore_names.join(', ')} for Storage policy #{vm_type.storage_policy_name}")
+        return "^(#{datastore_names.map { |name| Regexp.escape(name) }.join('|')})$", vm_type.storage_policy_name
+      end
+
+      # 2. Check if vm_type datastore or SDRS enabled datastore cluster exists
+      # return if exists
       datastore_names = vm_type.datastore_names
       unless vm_type.datastore_clusters.empty?
         sdrs_enabled_datastore_clusters = vm_type.datastore_clusters.select(&:drs_enabled?)
         datastores = sdrs_enabled_datastore_clusters.map { |datastore_cluster| datastore_cluster.datastores }.flatten
         datastore_names.concat(datastores.map(&:name))
       end
       datastore_names = datastore_names.compact
-      if datastore_names.empty? && vm_type.datastore_clusters.empty?
-        logger.info("Using global ephemeral disk datastore pattern: #{vm_type.datacenter.ephemeral_pattern}")
-        vm_type.datacenter.ephemeral_pattern
-      else
+      # @TA: TODO Do we need second condition here?
+      if !(datastore_names.empty? && vm_type.datastore_clusters.empty?)
```

Comment:
I remember adding this as 1 of the scenarios failed. This is for the scenario where you specify datastore_cluster and dont find any datastores. You want to default to global pattern only when datastores/datastore_clusters are not defined at vm_type level